src/anomalib/metrics/per_image/pimo_numpy.py

```diff
@@ -0,0 +1,461 @@
+        )
+        raise RuntimeError(msg)
+
+    # limit the curves to the integration range [lbound, ubound]
+    shared_fpr_bounded: ndarray = shared_fpr[thresh_lower_bound_idx : (thresh_upper_bound_idx + 1)]
+    per_image_tprs_bounded: ndarray = per_image_tprs[:, thresh_lower_bound_idx : (thresh_upper_bound_idx + 1)]
+
+    # `shared_fpr` and `tprs` are in descending order; `flip()` reverts to ascending order
+    shared_fpr_bounded = np.flip(shared_fpr_bounded)
+    per_image_tprs_bounded = np.flip(per_image_tprs_bounded, axis=1)
+
+    # the log's base does not matter because it's a constant factor canceled by normalization factor
+    shared_fpr_bounded_log = np.log(shared_fpr_bounded)
+
+    # deal with edge cases
+    invalid_shared_fpr = ~np.isfinite(shared_fpr_bounded_log)
+
+    if invalid_shared_fpr.all():
+        msg = (
+            "Cannot compute AUPIMO because the shared fpr integration range is invalid). "
+            "Try increasing the number of thresholds."
+        )
+        raise RuntimeError(msg)
+
+    if invalid_shared_fpr.any():
+        msg = (
+            "Some values in the shared fpr integration range are nan. "
+            "The AUPIMO will be computed without these values."
+        )
+        warnings.warn(msg, RuntimeWarning, stacklevel=1)
+        logger.warning(msg)
+
+        # get rid of nan values by removing them from the integration range
+        shared_fpr_bounded_log = shared_fpr_bounded_log[~invalid_shared_fpr]
+        per_image_tprs_bounded = per_image_tprs_bounded[:, ~invalid_shared_fpr]
+
+    num_points_integral = int(shared_fpr_bounded_log.shape[0])
+
+    if num_points_integral <= 30:
+        msg = (
+            "Cannot compute AUPIMO because the shared fpr integration range doesnt have enough points. "
+            f"Found {num_points_integral} points in the integration range. "
+            "Try increasing `num_threshs`."
+        )
+        if not force:
+            raise RuntimeError(msg)
+        msg += " Computation was forced!"
+        warnings.warn(msg, RuntimeWarning, stacklevel=1)
+        logger.warning(msg)
+
+    if num_points_integral < 300:
+        msg = (
+            "The AUPIMO may be inaccurate because the shared fpr integration range doesnt have enough points. "
+            f"Found {num_points_integral} points in the integration range. "
+            "Try increasing `num_threshs`."
+        )
+        warnings.warn(msg, RuntimeWarning, stacklevel=1)
+        logger.warning(msg)
+
+    aucs: ndarray = np.trapz(per_image_tprs_bounded, x=shared_fpr_bounded_log, axis=1)
+
+    # normalize, then clip(0, 1) makes sure that the values are in [0, 1] in case of numerical errors
+    normalization_factor = aupimo_normalizing_factor(fpr_bounds)
+    aucs = (aucs / normalization_factor).clip(0, 1)
+
+    return threshs, shared_fpr, per_image_tprs, image_classes, aucs, num_points_integral
```

Comment:
@ashwinvaidya17 
> Do you think we should capture these in a dataclass?

there is a dataclass for the torch interface but not for the numpy interface
i couldnt figure out a nice, maintainable way to make the two versions
perhaps we can discuss this in a call? i'll explain better and maybe you'll know a good solution